### ISSUE TITLE
node position

### ISSUE THREAD

**Issue by cherob67** [hanatos/vkdt#70]
Good morning,

I would like to make a suggestion regarding the position of the nodes. I think it would be easier to add these positions to the .cfg file following the list of nodes, links between nodes and parameters.
When you make a backup of your photo (or video) collection, the advantage of this solution is to include all the information related to the saved photos.

Same thing for .pst files (presets) or default pipes (ex: default-darkroom.i-raw).

Another wish: when customizing the default pipe it would be better to save the modified file in the directory ~/.config/vkdt so as not to lose the modifications when reinstalling vkdt. The modified file is to be used as a priority thereafter.

One last question: I couldn't use the blocks. Are they definitely replaced by the presets?

Thanks !

**Comment by hanatos** [comment]
heya, let me try to answer that:

- node positions: are kinda gui stuff, so i was hesitant to put them into .cfg files. the backup argument is valid though, and arguably grabbing the files from a directory containing only cryptic hashes is not easy. so putting into cfg may not be a bad idea, have to think about it.
- by now all the files should be read first from `~/.config/vkdt` and only after that in the system wide install directory. that should work for your default pipelines, colour profiles, presets, and blocks too.
- i certainly don't want to put presets/blocks into cfg files. that's not what you meant?
- blocks: since i deleted the `pipeline config` tab in darkroom mode, the only way of adding them is to select a module on the graph, then click the `insert block before this` button in the right panel of the graph editor:
![2023-03-19-103757_742x230_scrot](https://user-images.githubusercontent.com/1574185/226166409-8c024732-a894-4402-a584-ac848851398a.png)
which, after disentangling the node positions a bit, results in this (i.e. works here):
![2023-03-19-103733_797x838_scrot](https://user-images.githubusercontent.com/1574185/226166412-10b04c1c-82dd-47af-ae30-a56ae68a50a9.png)
what is the issue on your end? does it not work? i used the `draw` block, i.e. a shipped builtin.

**Comment by cherob67** [comment]
Thanks for your answer.

directory ~/.config/vkdt : oh ! i don't think it was possible for the default pipes. I try and it works fine.

blocks: yes it works:  the reason was i had tried with ctrl-b and this didn't work.

presets and blocks file: I just think to be able to add the position of the nodes that compose them in the corresponding pst or cfg files. For now, in the node editor,  when adding a preset or a block, the nodes appear piled up in a corner of the screen.

**Comment by hanatos** [comment]
okay i see what you mean. that makes a lot of sense. probably have to think about a way forward how to store gui stuff. i like the idea that presets/blocks bring their gui portion too. and that you can open a .cfg like a project file, similar to blender: stores configuration of gui too, might include window arrangement etc in the future (imgui docking style, dual monitor/popped out window). maybe it could have a "gui only" section that can safely be skipped.

**Comment by hanatos** [comment]
yes such an implicit initial ordering is in place in this line of code:
https://github.com/hanatos/vkdt/blob/master/src/gui/render_nodes.cc#L402

it just goes from left to right through the module list.

the next branch inits from positions that have been read from file.

probably the thing with presets and blocks is that they have to be merged into the current graph, i.e. potentially move around original graph nodes to make space and then move the preset/block stuff to the free spot. maybe a simple thing like selecting the bunch after inserting could help move them around manually too.

**Comment by cherob67** [comment]
Good morning

I want to share some thoughts regarding blocks and presets. I started from a "life-size" test of retouching on a personal photo that I have already retouched in DT.
Here is the photo (jpeg associated with the raw):
![_MG_6873-a](https://user-images.githubusercontent.com/46007178/227621470-0c0063c9-6a57-410a-8f30-f5f09e06de6d.JPG)

Here is the principle of treatment by zones with modification (according to the zones) of the luminosity, the color, the contrast and the local contrast
![_MG_6873_zones_a](https://user-images.githubusercontent.com/46007178/227621365-02b82ffa-57a3-4ced-b7a9-379a742fcf40.jpg)

Here is the photo obtained with VKDT
![_MG_6873 CR2_01_0000-a](https://user-images.githubusercontent.com/46007178/227621574-83bb34e1-7983-4c9b-9d2f-32bffa6eb734.jpg)

And here is the the pipe
![pipe_ex](https://user-images.githubusercontent.com/46007178/227621704-df75e186-44fc-4e80-ac90-bb87d0720406.jpg)

which I propose the following structure:
![pipe_ex_a](https://user-images.githubusercontent.com/46007178/227621816-77eec906-2b23-4f83-a88d-752ddba7704d.jpg)

From this structuring, the idea proposed is to establish a level above the elementary modules which would be the elements (name to be found - perhaps the blocks or presets) corresponding of course to an aggregation of elementary modules and ensuring more complex and complete functions. These blocks/presets could be established by the developers with the possibility for each user to create (and share) their own.
![pipe_ex_b](https://user-images.githubusercontent.com/46007178/227621885-13f7f75c-32c2-4449-b865-4dd18028366d.jpg)

The module editor would provide access to this intermediate level allowing easier development and reading of the pipe. The settings corresponding to each module composing an element would be accessible on the right column with an indentation to distinguish the different modules.

**Comment by hanatos** [comment]
the cfg file format now includes module positions as (optional) first class citizens right in the line of defining the module. i think this closes this particular issue, the other two related ones should cover the other aspects.
